clippy_lints/src/exit.rs

```diff
@@ -0,0 +1,62 @@
+use crate::utils::{match_def_path, paths, qpath_res, span_lint};
+use if_chain::if_chain;
+use rustc::hir::{Expr, ExprKind, Item, ItemKind, Node};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** `exit()`  terminates the program and doesn't provide a
+    /// stack trace.
+    ///
+    /// **Why is this bad?** Ideally a program is terminated by finishing
+    /// the main function.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```ignore
+    /// std::process::exit(0)
+    /// ```
+    pub EXIT,
+    restriction,
+    "`std::process::exit` is called, terminating the program"
+}
+
+declare_lint_pass!(Exit => [EXIT]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Exit {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, e: &'tcx Expr) {
+        if_chain! {
+            if let ExprKind::Call(ref path_expr, ref _args) = e.kind;
+            if let ExprKind::Path(ref path) = path_expr.kind;
+            if let Some(def_id) = qpath_res(cx, path, path_expr.hir_id).opt_def_id();
+            if match_def_path(cx, def_id, &paths::EXIT);
+            then {
+                let mut parent = cx.tcx.hir().get_parent_item(e.hir_id);
+                // We have to traverse the parents upwards until we find a function
+                // otherwise a exit in a let or if in main would still trigger this
```

Comment:
Have you tested this? `let`s and `if`s are expressions, not items. See [`ItemKind`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/enum.ItemKind.html) on possible items `get_parent_item` should find.

Reply:
I did some research, it seems get_parent_item will skip non items, which is fine as we don't fare for anything but the first function parent, so if there is a `if ... { exit(0) }` it will still just skip the if and right go to the function.

That said I'm running in a odd problem and I'm a bit stuck. I added another test to ensure this works with a if un a function, and here it gets odd. Even so the `span_lint` is hit for both only one error is ever emitted. So I'm a bit stomped. I left the test failing.

A hand with why only one span is emitted would be greatly appreciated.

Reply:
> I did some research, it seems get_parent_item will skip non items

So the loop shouldn't be necessary? The loop will only help, when `exit` is used in a nested function inside main. IMO we should still lint this though. We only want to bail out, if `exit` is directly used inside `main`

Reply:
> Even so the span_lint is hit for both only one error is ever emitted

That is really weird.

Reply:
Good point, I removed the loop, I had the thought that there might something else in the path we need skip but you're right the next item should be the function I removed the loop!

And yes that's really weird o.O